revocation/revocation.go

```diff
@@ -0,0 +1,249 @@
+				return CheckOCSPError{Err: errors.New("invalid chain: expected chain to be correct and complete with each cert issued by the next in the chain")}
+			}
+			r.wg.Add(1)
+			// Assume cert chain is accurate and next cert in chain is the issuer
+			go r.certOCSPStatus(cert, certChain[i+1], &certResults, i)
+		} else {
+			if !isCorrectIssuer(cert, cert) {
+				return CheckOCSPError{Err: errors.New("invalid chain: expected chain to end with root cert")}
+			}
+			// Last is root cert, which will never be revoked by OCSP
+			certResults[len(certChain)-1] = nil
+		}
+	}
+
+	r.wg.Wait()
+	return r.mergeErrorsFunc(certResults)
+}
+
+func isCorrectIssuer(subject *x509.Certificate, issuer *x509.Certificate) bool {
+	if err := subject.CheckSignatureFrom(issuer); err != nil {
+		return false
+	}
+	if !bytes.Equal(issuer.RawSubject, subject.RawIssuer) {
+		return false
+	}
+	return true
+}
+
+func (r *revocation) certOCSPStatus(cert *x509.Certificate, issuer *x509.Certificate, results *[]error, resultIndex int) error {
+	defer r.wg.Done()
+	var err error
+
+	ocspURLs := cert.OCSPServer
+	if len(ocspURLs) == 0 {
+		// OCSP not enabled for this certificate.
+		(*results)[resultIndex] = NoOCSPServerError{}
+		return NoOCSPServerError{}
+	}
+
+	serverErrs := make([]error, len(ocspURLs))
+	for serverIndex, server := range ocspURLs {
+		resp, err := r.ocspRequest(cert, issuer, server)
+		if err != nil {
+			// If there is a server error, attempt all servers before determining what to return to the user
+			serverErrs[serverIndex] = err
+			continue
+		}
+
+		if time.Now().After(resp.NextUpdate) {
+			err = errors.New("expired OCSP response")
+			serverErrs[serverIndex] = err
+			continue
+		}
+
+		foundNoCheck := false
+		pkixNoCheckOID := "1.3.6.1.5.5.7.48.1.5"
+		for _, extension := range resp.Extensions {
+			if !foundNoCheck && extension.Id.String() == pkixNoCheckOID {
+				foundNoCheck = true
+			}
+		}
+		if !foundNoCheck {
+			// This will be ignored until CRL is implemented
+			// If it isn't found, CRL should be used to verify the OCSP response
+			fmt.Printf("\n[WARNING] An ocsp signing cert is missing the id-pkix-ocsp-nocheck extension (%s)\n", pkixNoCheckOID)
+		}
+
+		if resp.Status == ocsp.Revoked {
+			if time.Now().After(resp.RevokedAt) {
+				(*results)[resultIndex] = RevokedInOCSPError{}
+				return RevokedInOCSPError{}
+			} else {
+				(*results)[resultIndex] = nil
+				return nil
+			}
+		} else if resp.Status == ocsp.Good {
+			(*results)[resultIndex] = nil
+			return nil
+		} else {
+			(*results)[resultIndex] = UnknownInOCSPError{}
+			return UnknownInOCSPError{}
+		}
+
+	}
+	// Errors in all server responses, determine the most pressing
+	err = r.mergeErrorsFunc(serverErrs)
+	(*results)[resultIndex] = err
+	return err
+}
+
+func (r *revocation) ocspRequest(cert, issuer *x509.Certificate, server string) (*ocsp.Response, error) {
+	ocspRequest, err := ocsp.CreateRequest(cert, issuer, &ocsp.RequestOptions{Hash: crypto.SHA1})
```

Comment:
Can the hash here be SHA256 or something else?

Reply:
Resolving. Further discussion on Hash to use tracked [here](https://github.com/notaryproject/notation-core-go/pull/134#discussion_r1152453847)